server/src/completions.ts

```diff
@@ -0,0 +1,563 @@
+}
+
+export function getSuggestionsForAttributes(
+  blockType: string,
+  position: Position,
+  document: TextDocument,
+): CompletionList {
+  const symbolBeforePosition = document.getText({
+    start: { line: position.line, character: position.character - 2 },
+    end: { line: position.line, character: position.character },
+  })
+  let suggestions: CompletionItem[] = []
+
+  if (symbolBeforePosition === '@@') {
+    suggestions = getSuggestionForBlockAttribute(blockType)
+  } else if (symbolBeforePosition === ' @') {
+    const blockAttributeSuggestions: CompletionItem[] = getSuggestionForBlockAttribute(
+      blockType,
+    )
+    for (let i = 0; i < blockAttributeSuggestions.length; i++) {
+      blockAttributeSuggestions[i].data =
+        '@' + blockAttributeSuggestions[i].label
+    }
+    suggestions = getSuggestionForFieldAttribute(
+      blockType,
+      position,
+      document,
+    ).concat(blockAttributeSuggestions)
+  } else {
+    // valid schema
+    const fieldAttributeSuggestions = getSuggestionForFieldAttribute(
+      blockType,
+      position,
+      document,
+    )
+    for (let i = 0; i < fieldAttributeSuggestions.length; i++) {
+      fieldAttributeSuggestions[i].label =
+        '@' + fieldAttributeSuggestions[i].label
+    }
+    const blockAttributeSuggestions = getSuggestionForBlockAttribute(blockType)
+    for (let i = 0; i < blockAttributeSuggestions.length; i++) {
+      blockAttributeSuggestions[i].label =
+        '@@' + blockAttributeSuggestions[i].label
+    }
+    suggestions = fieldAttributeSuggestions.concat(blockAttributeSuggestions)
+  }
+
+  return {
+    items: suggestions,
+    isIncomplete: false,
+  }
+}
+
+export function getAllRelationNamesWithoutAst(
+  document: TextDocument,
+  currentModelStartLine: number,
+): Array<string> {
+  const modelNames: Array<string> = []
+  for (let i = 0; i < document.lineCount; i++) {
+    if (currentModelStartLine === i) {
+      // do not suggest the model name we are currently in
+      continue
+    }
+    const currentLine = getCurrentLine(document, i)
+    if (
+      (currentLine.includes('model') || currentLine.includes('enum')) &&
+      currentLine.includes('{')
+    ) {
+      // found a block
+      const trimmedLine = currentLine.trim()
+      const blockType = currentLine.trim().replace(/ .*/, '')
+      const blockName = trimmedLine
+        .substring(blockType.length, trimmedLine.length - 1)
+        .trim()
+
+      modelNames.push(blockName)
+      // block is at least 2 lines long
+      i++
+    }
+  }
+  return modelNames
+}
+
+export function getSuggestionsForTypes(
+  foundBlock: MyBlock,
+  document: TextDocument,
+): CompletionList {
+  let suggestions: CompletionItem[] = corePrimitiveTypes
+
+  if (foundBlock instanceof MyBlock) {
+    // get all model names
+    const modelNames: Array<string> = getAllRelationNamesWithoutAst(
+      document,
+      foundBlock.start.line,
+    )
+    suggestions = suggestions.concat(
```

Comment:
A faster alternative is:
```ts
suggestions.push(...toCompletionItems(modelNames, CompletionItemKind.TypeParameter))
```
https://dev.to/uilicious/javascript-array-push-is-945x-faster-than-array-concat-1oki